staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go

```diff
@@ -0,0 +1,322 @@
+	// definition. A single TotalCost is allocated for each CustomResourceDefinition and passed through the stack as the
+	// CustomResourceDefinition's OpenAPIv3 schema is recursively validated.
+	TotalCost *TotalCost
+}
+
+// HasValidationRules returns true if the current schema node contains x-kubernetes-validations rules.
+func (c *CELSchemaContext) HasValidationRules() bool {
+	return len(c.jsonSchema.XValidations) > 0
+}
+
+// TypeInfo returns the CELTypeInfo for this CELSchemaContext node.  Returns nil if current level or above does not have
+// x-kubernetes-validations rules.
+func (c *CELSchemaContext) TypeInfo() *CELTypeInfo {
+	if !c.withinValidationRuleScope {
+		return nil
+	}
+	if c.typeInfo != nil {
+		return c.typeInfo
+	}
+
+	// If able to get the type info from the parent's type info, prefer this approach
+	// since it is more efficient.
+	if c.parent != nil {
+		parentTypeInfo := c.parent.TypeInfo()
+		if parentTypeInfo != nil && c.typeInfoAccessor != nil {
+			c.typeInfo = c.typeInfoAccessor.accessTypeInfo(parentTypeInfo)
+			return c.typeInfo
+		}
+	}
+	// If unable to get the type info from the parent, convert the jsonSchema to type info.
+	// This is expensive for large schemas.
+	typeInfo, err := c.converter(c.jsonSchema, c.parent == nil || c.jsonSchema.XEmbeddedResource)
+	if err == nil {
+		c.typeInfo = typeInfo
+	}
+	return c.typeInfo
+}
+
+// CELTypeInfo represents all the typeInfo needed by CEL to compile x-kubernetes-validations rules for a schema node.
+type CELTypeInfo struct {
+	// Schema is a structural schema for this CELSchemaContext node. It is null if the current schema
+	// has no x-kubernetes-validations rules at or above this level or if the schema is not a valid structural schema.
+	Schema *structuralschema.Structural
+	// DeclType is a CEL declaration representation of Schema of this CELSchemaContext node. It is null if the current schema
+	// has no x-kubernetes-validations rules at or above this level or if the schema is not a valid structural schema.
+	DeclType *model.DeclType
+}
+
+// Converter converts from JSON schema to a structural schema and a CEL declType, or returns an error if the conversion
+// fails.
+type Converter func(schema *apiextensions.JSONSchemaProps, isRoot bool) (*CELTypeInfo, error)
+
+// DefaultConverter converts using standard conversion utilities.
+func DefaultConverter(schema *apiextensions.JSONSchemaProps, isRoot bool) (*CELTypeInfo, error) {
+	structural, err := structuralschema.NewStructural(schema)
+	if err != nil {
+		return nil, err
+	}
+	declType := model.SchemaDeclType(structural, isRoot)
+	if declType == nil {
+		return nil, fmt.Errorf("unable to convert structural schema to CEL declarations")
+	}
+	return &CELTypeInfo{structural, declType}, nil
+}
+
+// RootCelContext constructs CELSchemaContext for the given root schema.
+func RootCelContext(schema *apiextensions.JSONSchemaProps, converter Converter) CELSchemaContext {
```

Comment:
if the `converter` param is only for wrapping `DefaultConverter` for unit testing, and the only testing that is using it is in-package, I'd suggest:
1. document the field as being for unit test purposes and set to DefaultConverter otherwise
2. omitting the param from this constructor
3. initializing the field to `DefaultConverter` in this function
4. make `DefaultConverter` private as well, since the only callers are in-package
5. privately stomping the `converter` field after calling `RootCelContext` in validation/validation_test.go

that keeps callers from having to worry about this bit and reduces the surface area

Reply:
Nice, yeah, that cleans up the surface area.